.circleci/config.yml

```diff
@@ -10,61 +10,26 @@ jobs:
     steps:
       - checkout
 
-      # Factoring demo
-#      - run: &make-virtualenv
-#          command: |
-#            python -m virtualenv env
-#
-#      - run:
-#          command: |
-#            . env/bin/activate
-#            pip install -r factoring/requirements_tests.txt
-#
-#TODO: add caching back in, maybe use loop
-      - run:
+      #TODO: add caching once demos share the same requirements file and can live in the same venv
+      - run: &install-and-test-template
+          name: install dependences and run tests
           command: |
-            demos=$(ls -d */ | grep -v "tests" | grep -v "circuit-fault-diagnosis" | sed 's/\///')
-            for demo in $demos
-            do
-                env="${demo}_env"
-                python -m virtualenv $env
-                . $env/bin/activate
-                pip install -r $demo/requirements_tests.txt
+            shopt -s extglob
+
+            for demo in !(tests|circuit-fault-diagnosis); do
+              if [ -d "$demo" ]; then
+                env="env_$demo"
+
+                python -m virtualenv "$env"
+                . "$env/bin/activate"
+
+                pip install -r "$demo/requirements_tests.txt"
                 coverage run -m unittest discover
+
                 deactivate
+              fi
```

Comment:
The code looks clean and correct now.

I would just use this opportunity to note it could be simplified a bit by avoiding venv activation, and using the virtual environment executables explicitly:
```
for demo in !(tests|circuit-fault-diagnosis); do
    if [[ -d "$demo" ]]; then
        python -m virtualenv "env/$demo"
        "env/$demo/bin/pip" install -r "$demo/requirements_tests.txt"
        "env/$demo/bin/coverage" run -m unittest discover
    fi
done
```

Also note I put all `$demo` dirs under a single (repo-level) `env` dir.

Reply:
I had initially kept them in separate environments because I was hoping to cache the environments. However, since caching will be added once the demos share the same requirements, I will store all `$demo` dirs under `env`.

I will apply the code you have suggested (i.e. without activation)

Reply:
Actually, if it's not a performance issue, I would prefer to keep the virtualenv activation. The code seems more readable to me if we simply use
```
. "env/$demo/activate"
pip install -r "$demo/requirements_tests.txt"
coverage run -m unittest discover
```
rather than
```
 "env/$demo/bin/pip" install -r "$demo/requirements_tests.txt"
 "env/$demo/bin/coverage" run -m unittest discover
```

Reply:
I like storing all the virtual environments in one dir, `env`, so I'll add that in.

Reply:
That’s fine. No performance difference. Shorter variant makes more sense when there’s only one call. Here it’s less readable, I agree.